clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -634,7 +698,7 @@ void abortUndrainedBatches(RuntimeException reason) {
             Deque<ProducerBatch> dq = getDeque(batch.topicPartition);
             boolean aborted = false;
             synchronized (dq) {
-                if (!batch.isClosed()) {
+                if ((transactionManager != null && !batch.hasSequence()) || (transactionManager == null && !batch.isClosed())) {
```

Comment:
Hmm.. seems we should always check `!batch.isClosed()`?

Reply:
This is related to your other point about the state of the records builder. Right now, `isClosed` is just `builtRecords != null`. But we do reopen the batch (ie. nullify the `builtRecords`) when we have to change the sequence numbers due to a fatal failure. That is why I had to modify this check.

Reply:
Since this function is for aborting undrained batches, maybe an alternative is to have an `isUndrained` method on `ProducerBatch` which checks if `drainedMs` is greater than 0?